tokio-util/src/io/utf8_reader.rs

```diff
@@ -0,0 +1,284 @@
+    /// assert_eq!(bytes_read, 8);
+    /// assert_eq!(str::from_utf8(&buffer[..bytes_read])?, "😀😬");
+    ///
+    /// // ... later on ...
+    ///
+    /// // `reader` now gives us the leftover byte from last time because the
+    /// // rest of the bytes that make up the third emoji have been supplied
+    /// // from the inner reader, as well as the fourth and fifth emojis. We
+    /// // have to slice `buffer` by `bytes_read` in order to not overwrite what
+    /// // was read in the previous read
+    /// let bytes_read_2 = reader.read(&mut buffer[bytes_read..]).await?;
+    /// assert_eq!(bytes_read_2, 12);
+    /// assert_eq!(str::from_utf8(&buffer[bytes_read..][..bytes_read_2])?, "😁😂😃");
+    /// assert_eq!(str::from_utf8(&buffer[..(bytes_read + bytes_read_2)])?, "😀😬😁😂😃");
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[derive(Debug, Clone)]
+    pub struct Utf8Reader<R> {
+        #[pin]
+        inner: R,
+        scrap_state: ScrapState,
+    }
+}
+
+impl<R> Utf8Reader<R> {
+    /// Create a new `Utf8Reader` from an underlying reader.
+    ///
+    /// Note that the generic parameter `R` is not constrained by either
+    /// [`AsyncRead`] or [`AsyncBufRead`], but in order to be able to use a
+    /// `Utf8Reader` as one (in order for `Utf8Reader` to implement
+    /// [`AsyncRead`] and/or [`AsyncBufRead`]), `R` must implement one or both
+    /// of those traits.
+    ///
+    /// [`AsyncRead`]: https://docs.rs/tokio/latest/tokio/io/trait.AsyncRead.html
+    /// [`AsyncBufRead`]: https://docs.rs/tokio/latest/tokio/io/trait.AsyncBufRead.html
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio_util::io::Utf8Reader;
+    ///
+    /// # const INNER_READER: () = ();
+    /// let utf8_reader = Utf8Reader::new(INNER_READER);
+    /// ```
+    pub const fn new(inner: R) -> Self {
+        Self {
+            inner,
+            scrap_state: ScrapState::NoScrap,
+        }
+    }
+
+    /// Get the inner reader out of the `Utf8Reader`, destroying the
+    /// `Utf8Reader` in the process.
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio_util::io::Utf8Reader;
+    ///
+    /// # const INNER_READER: () = ();
+    /// let utf8_reader = Utf8Reader::new(INNER_READER);
+    ///
+    /// // ... later on ...
+    /// let inner_reader = utf8_reader.into_inner();
+    /// ```
+    pub fn into_inner(self) -> R {
+        self.inner
+    }
+}
+
+impl<R> From<R> for Utf8Reader<R> {
+    fn from(inner: R) -> Self {
+        Self::new(inner)
+    }
+}
+
+impl<R: AsyncRead> AsyncRead for Utf8Reader<R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let me = self.as_mut().project();
+        match me.scrap_state {
+            ScrapState::Scrap(scrap, len) => {
+                buf.put_slice(&scrap[..*len]);
+                *me.scrap_state = ScrapState::NoScrap;
+                self.poll_read(cx, buf)
+            }
```

Comment:
This is incorrect if the provided `ReadBuf` is shorter than the scrap state.

Reply:
In that case, if there's no space left, should we return `ErrorKind::UnexpectedEof`? We shouldn't put part of the scrap in, as that'd break the invariant that `Utf8Reader` doesn't yield partial UTF-8 sequences.

Reply:
The only correct behavior here is to return the partial UTF-8 sequence. The guarantee would have to be changed to allow it.

Reply:
Is this correct?
```rust
ScrapState::Scrap(scrap, len) => {
    let remaining = buf.remaining();
    buf.put_slice(&scrap[..std::cmp::min(*len, remaining)]);
    *me.scrap_state = if remaining < *len {
        // will not have a len of zero, as remaining < len
        ScrapState::Scrap(create_scrap(&scrap[remaining..*len]), *len - remaining)
    } else {
        ScrapState::NoScrap
    };

    self.poll_read(cx, buf)
}
```

Reply:
When you have successfully read something, you should just return `Poll::Ready(Ok(()))`. Otherwise you risk the `poll_read` call returns `Poll::Pending`, and you aren't supposed to have read anything if you're returning that.

Reply:
This test passes, so I think so
```rust
#[tokio::test]
#[allow(const_item_mutation)]
async fn too_short_read_buf() {
    let mut buf = [0; 3];
    assert_eq!(assert_ok!(Pin::new(&mut EMOJIS).read(&mut buf).await), 3);
    assert_eq!(buf, &EMOJIS[..3]);
}
```

Reply:
Oh sorry, I just saw that. Does that mean I shouldn't read anything into `buf` if there is a scrap? If there's a scrap left over, should I always return `Poll::Ready(Ok(()))` after putting it onto `buf`?

Reply:
I was thinking, and here are the scenarios I saw:
* I put the scrap into `buf` and `return Poll::Ready(Ok(()))`. The user sees a partial UTF-8 character
* The same as last one, but this time I `set_filled` to the valid UTF-8. This makes it look like 0 bytes have been read, and thus we have reached EOF
* I put the scrap into `buf`, then immediately call the inner `poll_read`. If the inner `poll_read` returns `Poll::Pending`, I've broken that invariant

None of those are good.

However, I realized that I could match on the inner reader. If it's `Poll::Pending`, I undo the scrap stuff and return `Poll::Pending`. If it's `Poll::Ready`, then I do the usual stuff. Could that work?